### PR TITLE
ARROW-4205: [Gandiva] Support for decimal multiply

### PR DESCRIPTION
**Pull request by pravindra**
- fastpath : safe multiply the two values (when result precision is < 38)
- if large values, convert to 256-bit (boost), multiply and then, scale down
- avoid converting to 256-bit wherever possible
- track overflow (TODO: generate errors on overflow)

### CODE REVIEW

File: cpp/src/gandiva/function_registry_string.cc [hunk 1]
```diff
@@ -38,6 +41,13 @@ std::vector<NativeFunction> GetStringFunctionRegistry() {
       BINARY_RELATIONAL_SAFE_NULL_IF_NULL_UTF8_FN(starts_with),
       BINARY_RELATIONAL_SAFE_NULL_IF_NULL_UTF8_FN(ends_with),
 
+      UNARY_OCTET_LEN_FN(octet_length),
```
Comment (pitrou) [inline cpp/src/gandiva/function_registry_string.cc]: Are these additions related to decimal multiplication?

File: cpp/src/gandiva/decimal_xlarge.cc [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  args = {types->i64_type(),      // int64_t x_high
+          types->i64_type(),      // uint64_t x_low
+          types->i64_type(),      // int64_t y_high
+          types->i64_type(),      // uint64_t x_low
+          types->i32_type(),      // int32_t reduce_scale_by
+          types->i64_ptr_type(),  // int64_t* out_high
+          types->i64_ptr_type(),  // uint64_t* out_low
+          types->i8_ptr_type()};  // bool* overflow
+
+  engine->AddGlobalMappingForFunc("gdv_multiply_and_scale_down",
```
Comment (pitrou) [inline cpp/src/gandiva/decimal_xlarge.cc]: What is the purpose of this declaration? AFAICT, `gdv_multiply_and_scale_down` is only called from `MultiplyMaxPrecision` which is a regular C++ function.

I find the current call path quite convoluted, and it's not obvious what will be compiled to LLVM bitcode and what will be compiled to native code ahead of time.

**Review by pitrou — COMMENTED** [review]
A couple comments and questions. I find the code flow a bit difficult to follow...

File: cpp/src/gandiva/function_registry_string.cc (hunk 1 above)
Comment (pravindra) [reply cpp/src/gandiva/function_registry_string.cc]: no - minor cleanup. I moved them from function_registry_arithmetic.cc.

File: cpp/src/gandiva/decimal_xlarge.cc (hunk 2 above)
Comment (pravindra) [reply cpp/src/gandiva/decimal_xlarge.cc]: only files in the folder src/gandiva/precompiled and basic_decimal.cc are compiled to llvm bitcode ahead of time. The rest are compiled to native code.

In the files that are precompiled to llvm bitcode, I'm avoiding references to libstdc++/boost since it complicates symbol resolution at run-time. In such cases, I add a C wrapper in non-precompiled-to-llvm code (i.e outside src/gandiva/precompiled), and export it as a global function in JIT.

I'll change the name of this to gdv_xlarge_multiply_and_scale_down and add a comment here on why it can't be precompiled-to-bitcode.

**Comment by pitrou** [comment]
But if `multiply_internal_decimal128_decimal128` (which is the outer function) is the only thing precompiled to IR, why is there another "C" function (`gdv_xlarge_multiply_and_scale_down`) in the middle of the call stack?

**Comment by pitrou** [comment]
Ah, sorry, I missed the comment :-(

File: cpp/src/gandiva/decimal_ir.cc [hunk 3]
```diff
@@ -353,6 +353,60 @@ Status DecimalIR::BuildSubtract() {
...
+  auto arg_iter = function->arg_begin();
+  ValueFull x(&arg_iter[0], &arg_iter[1], &arg_iter[2]);
+  ValueFull y(&arg_iter[3], &arg_iter[4], &arg_iter[5]);
+  ValueFull out(nullptr, &arg_iter[6], &arg_iter[7]);
+
+  auto entry = llvm::BasicBlock::Create(*context(), "entry", function);
+  ir_builder()->SetInsertPoint(entry);
+
+  // Make call to pre-compiled IR function.
+  // TODO: add fast paths.
```
Comment (pitrou) [inline cpp/src/gandiva/decimal_ir.cc]: Not sure I understand this comment. There is already a fast path, no? The only issue might be that `gandiva::decimalops::Multiply` isn't inlined into its caller.

Comment (pravindra) [reply cpp/src/gandiva/decimal_ir.cc]: I wanted to try the intrinsics-with-overflow instructions. I've removed the comment - will experiment with this after adding a micro benchmark.

**Comment by pravindra** [comment]
@pitrou I've added comments, renamed fns to address the issues you raised. Can you please take a look and see if it looks good now ? Please let me know if you want me to rework anything in this patch.

File: ci/travis_script_manylinux.sh [hunk 4]
```diff
@@ -53,7 +53,7 @@ for PYTHON_TUPLE in ${PYTHON_VERSIONS}; do
     -e UNICODE_WIDTH=$UNICODE_WIDTH \
     -v $PWD:/io \
     -v $PWD/../../:/arrow \
-    quay.io/xhochy/arrow_manylinux1_x86_64_base:latest \
+    quay.io/pravindra/arrow_manylinux1_x86_64_base:latest \
```
Comment (pravindra) [inline ci/travis_script_manylinux.sh]: @xhochy unfortunately, there was a typo in the change I gave you for the boost change in build_boost.sh (I had put multi_precision instead of multiprecision).

I fixed that, and added a trigger in quay.io to build. Please let me know if that's okay. I've included the change to build_boost.sh (fixing the typo) as part of this PR.

Comment (xhochy) [reply ci/travis_script_manylinux.sh]: 👍

**Review by pitrou — APPROVED** [review]
+1. @pravindra can you merge?

**Comment by pravindra** [comment]
thanks @pitrou and @xhochy. I'll do the merge,
